server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java

```diff
@@ -32,10 +33,30 @@
 import reactor.core.publisher.Signal;
 import reactor.core.publisher.SynchronousSink;
 import reactor.util.context.Context;
+import reactor.util.function.Tuple2;
 
 public class ReactorUtils {
 
     public static final String MDC_KEY_PREFIX = "MDC-";
+    private static final Duration DELAY = Duration.ZERO;
+
+    /**
+     * Ensures windowMaxSize gets processed every windowDuration.
+     *
+     * Caller needs to limit concurrency of subsequent running operations to avoid overwhelming the system in case single
+     * item processing is longer than windowDuration.
+     *
+     * Example:
+     *
+     * int itemsPerWindow = 20;
+     * throttle(flux, Duration.ofSeconds(1), itemsPerWindow)
+     *     .flatMap(operation, itemsPerWindow)
+     */
+    public static <T> Flux<T> throttle(Flux<T> flux, Duration windowDuration, int windowMaxSize) {
+        return flux.windowTimeout(windowMaxSize, windowDuration)
+            .zipWith(Flux.interval(DELAY, windowDuration))
```

Comment:
how is the backpressure handled ?

Reply:
> how is the backpressure handled ?

Combinaison of windowing, and rate limiting (if we can call it "back-pressure").